File: wei-ui/src/api/knowledge.ts
/**
 * Knowledge Base API client
 * 
 * Endpoints:
 * - GET /proxy/admin/knowledge/documents - List documents
 * - POST /proxy/admin/knowledge/documents - Create document
 * - GET /proxy/admin/knowledge/documents/:id - Get document
 * - PUT /proxy/admin/knowledge/documents/:id - Update document
 * - DELETE /proxy/admin/knowledge/documents/:id - Delete document
 * - POST /proxy/admin/knowledge/documents/upload - Upload file
 * - POST /proxy/admin/knowledge/documents/:id/extract - Extract knowledge
 * - GET /proxy/admin/knowledge/documents/:id/download - Download file
 * - POST /proxy/admin/knowledge/query - Query knowledge base
 * - GET /proxy/admin/knowledge/stats - Get statistics
 */

import { request } from './request';

export type DocumentStatus = 'pending' | 'extracting' | 'extracted' | 'processing' | 'processed' | 'failed';
export type DocumentType = 'text' | 'pdf' | 'doc' | 'image' | 'video' | 'audio' | 'rss' | 'web';

export interface KnowledgeChunk {
  id?: string;
  content: string;
  document_id: string;
  model_id?: string;
  topic?: string;
  metadata: Record<string, any>;
  score?: number;
  created_at: string;
}

export interface Document {
  id?: string;
  title: string;
  content?: string;
  content_type: DocumentType;
  file_path?: string;
  file_size?: number;
  mime_type?: string;
  status: DocumentStatus;
  status_message?: string;
  model_id?: string;
  shared: boolean;
  topics: string[];
  extraction_confidence?: number;
  chunk_count: number;
  word_count?: number;
  source_url?: string;
  source_type?: string;
  tags: string[];
  custom_metadata: Record<string, any>;
  created_at: string;
  updated_at?: string;
}

export interface DocumentListParams {
  model_id?: string;
  status?: DocumentStatus;
  content_type?: DocumentType;
  shared?: boolean;
  search?: string;
  skip?: number;
  limit?: number;
}

export interface DocumentListResponse {
  total: number;
  documents: Document[];
  skip: number;
  limit: number;
}

export interface DocumentCreateRequest {
  title: string;
  content_type?: DocumentType;
  model_id?: string;
  shared?: boolean;
  tags?: string[];
  source_url?: string;
  source_type?: string;
  custom_metadata?: Record<string, any>;
}

export interface DocumentUpdateRequest {
  title?: string;
  content?: string;
  model_id?: string;
  shared?: boolean;
  tags?: string[];
  topics?: string[];
  custom_metadata?: Record<string, any>;
}

export interface FileUploadResponse {
  success: boolean;
  document_id?: string;
  file_name: string;
  file_size: number;
  content_type: DocumentType;
  message: string;
  status: DocumentStatus;
}

export interface KnowledgeQueryRequest {
  query: string;
  model_id?: string;
  topic?: string;
  limit?: number;
  threshold?: number;
}

export interface KnowledgeQueryResponse {
  query: string;
  results: KnowledgeChunk[];
  total_found: number;
  threshold_used: number;
}

export interface ExtractionRequest {
  document_id: string;
  model_id?: string;
  threshold?: number;
  auto_classify?: boolean;
}

export interface ExtractionResult {
  document_id: string;
  success: boolean;
  chunks_created: number;
  topics_detected: string[];
  confidence_score?: number;
  message: string;
  processing_time_ms?: number;
}

export interface DocumentStats {
  total_documents: number;
  by_status: Record<string, number>;
  by_type: Record<string, number>;
  by_model: Record<string, number>;
  total_chunks: number;
  total_words: number;
  avg_extraction_confidence?: number;
}

// List documents
export function getDocuments(params: DocumentListParams = {}): Promise<DocumentListResponse> {
  return request.get('/knowledge/documents', { params });
}

// Create document
export function createDocument(data: DocumentCreateRequest): Promise<Document> {
  return request.post('/knowledge/documents', data);
}

// Get document
export function getDocument(id: string): Promise<Document> {
  return request.get(`/knowledge/documents/${id}`);
}

// Update document
export function updateDocument(id: string, data: DocumentUpdateRequest): Promise<Document> {
  return request.put(`/knowledge/documents/${id}`, data);
}

// Delete document
export function deleteDocument(id: string): Promise<{ success: boolean; message: string }> {
  return request.delete(`/knowledge/documents/${id}`);
}

// Upload file
export function uploadFile(
  file: File,
  options: {
    model_id?: string;
    shared?: boolean;
    title?: string;
    tags?: string;
  } = {}
): Promise<FileUploadResponse> {
  const formData = new FormData();
  formData.append('file', file);
  
  if (options.model_id) formData.append('model_id', options.model_id);
  if (options.shared) formData.append('shared', String(options.shared));
  if (options.title) formData.append('title', options.title);
  if (options.tags) formData.append('tags', options.tags);
  
  return request.post('/knowledge/documents/upload', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
}

// Extract knowledge from document
export function extractDocument(
  id: string,
  data: Omit<ExtractionRequest, 'document_id'> = {}
): Promise<ExtractionResult> {
  return request.post(`/knowledge/documents/${id}/extract`, data);
}

// Reprocess document
export function reprocessDocument(
  id: string,
  data: Omit<ExtractionRequest, 'document_id'> = {}
): Promise<ExtractionResult> {
  return request.post(`/knowledge/documents/${id}/reprocess`, data);
}

// Download document file
export function downloadDocument(id: string): Promise<Blob> {
  return request.get(`/knowledge/documents/${id}/download`, {
    responseType: 'blob',
  });
}

// Query knowledge base
export function queryKnowledge(data: KnowledgeQueryRequest): Promise<KnowledgeQueryResponse> {
  return request.post('/knowledge/query', data);
}

// Get statistics
export function getKnowledgeStats(): Promise<DocumentStats> {
  return request.get('/knowledge/stats');
}

File: app/models/knowledge.py
"""
Knowledge base models for document management and extraction.

Features:
- Document metadata
- Extraction results
- Knowledge chunks for vector storage
- Query results
"""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import Field, field_validator

from models.base import BaseModel, TimestampMixin


class DocumentStatus(str, Enum):
    """Document processing status."""
    PENDING = "pending"  # Just uploaded
    EXTRACTING = "extracting"  # Text extraction in progress
    EXTRACTED = "extracted"  # Text extraction complete
    PROCESSING = "processing"  # Knowledge extraction in progress
    PROCESSED = "processed"  # Fully processed
    FAILED = "failed"  # Processing failed


class DocumentType(str, Enum):
    """Document content type."""
    TEXT = "text"  # Plain text, markdown, etc.
    PDF = "pdf"  # PDF files
    DOC = "doc"  # Word documents
    IMAGE = "image"  # Images with OCR
    VIDEO = "video"  # Video transcriptions
    AUDIO = "audio"  # Audio transcriptions
    RSS = "rss"  # RSS feed content
    WEB = "web"  # Web page content


class KnowledgeChunk(BaseModel):
    """Knowledge chunk for vector storage."""
    
    id: Optional[str] = None
    content: str = Field(..., description="Chunk text content")
    document_id: str = Field(..., description="Parent document ID")
    model_id: Optional[str] = Field(None, description="Associated virtual model")
    topic: Optional[str] = Field(None, description="Classified topic")
    embedding: Optional[List[float]] = Field(None, description="Vector embedding")
    metadata: Dict[str, Any] = Field(default_factory=dict)
    score: Optional[float] = Field(None, description="Similarity score for queries")
    created_at: datetime = Field(default_factory=datetime.utcnow)


class Document(BaseModel, TimestampMixin):
    """Document metadata model."""
    
    id: Optional[str] = Field(None, alias="_id")
    title: str = Field(..., min_length=1, max_length=500)
    content: Optional[str] = Field(None, description="Extracted text content")
    content_type: DocumentType = Field(default=DocumentType.TEXT)
    file_path: Optional[str] = Field(None, description="Original file path")
    file_size: Optional[int] = Field(None, description="File size in bytes")
    mime_type: Optional[str] = Field(None, description="MIME type")
    
    # Processing status
    status: DocumentStatus = Field(default=DocumentStatus.PENDING)
    status_message: Optional[str] = Field(None, description="Status details or error message")
    
    # Associations
    model_id: Optional[str] = Field(None, description="Associated virtual model")
    shared: bool = Field(default=False, description="Whether shared across models")
    topics: List[str] = Field(default_factory=list, description="Classified topics")
    
    # Extraction metadata
    extraction_confidence: Optional[float] = Field(None, description="Extraction quality score")
    chunk_count: int = Field(default=0, description="Number of knowledge chunks")
    word_count: Optional[int] = Field(None, description="Word count")
    
    # Source info
    source_url: Optional[str] = Field(None, description="Original URL if downloaded")
    source_type: Optional[str] = Field(None, description="Source type (upload/rss/web)")
    
    # User metadata
    tags: List[str] = Field(default_factory=list)
    custom_metadata: Dict[str, Any] = Field(default_factory=dict)


class DocumentCreateRequest(BaseModel):
    """Request to create a document record."""
    
    title: str = Field(..., min_length=1, max_length=500)
    content_type: DocumentType = Field(default=DocumentType.TEXT)
    model_id: Optional[str] = Field(None, description="Associate with specific model")
    shared: bool = Field(default=False)
    tags: List[str] = Field(default_factory=list)
    source_url: Optional[str] = None
    source_type: Optional[str] = None
    custom_metadata: Dict[str, Any] = Field(default_factory=dict)


class DocumentUpdateRequest(BaseModel):
    """Request to update document metadata."""
    
    title: Optional[str] = Field(None, max_length=500)
    content: Optional[str] = None
    model_id: Optional[str] = None
    shared: Optional[bool] = None
    tags: Optional[List[str]] = None
    topics: Optional[List[str]] = None
    custom_metadata: Optional[Dict[str, Any]] = None


class DocumentListRequest(BaseModel):
    """Request to list documents with filters."""
    
    model_id: Optional[str] = None
    status: Optional[DocumentStatus] = None
    content_type: Optional[DocumentType] = None
    shared: Optional[bool] = None
    topic: Optional[str] = None
    search: Optional[str] = None  # Text search in title/content
    tags: Optional[List[str]] = None
    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=20, ge=1, le=100)


class DocumentListResponse(BaseModel):
    """Response for document list."""
    
    total: int
    documents: List[Document]
    skip: int
    limit: int


class KnowledgeQueryRequest(BaseModel):
    """Request to query knowledge base."""
    
    query: str = Field(..., min_length=1, description="Search query text")
    model_id: Optional[str] = Field(None, description="Filter by model")
    topic: Optional[str] = Field(None, description="Filter by topic")
    limit: int = Field(default=5, ge=1, le=20, description="Max results")
    threshold: Optional[float] = Field(None, ge=0, le=1, description="Similarity threshold")
    
    @field_validator('threshold')
    def validate_threshold(cls, v):
        if v is not None and (v < 0 or v > 1):
            raise ValueError('Threshold must be between 0 and 1')
        return v


class KnowledgeQueryResponse(BaseModel):
    """Response for knowledge query."""
    
    query: str
    results: List[KnowledgeChunk]
    total_found: int
    threshold_used: float


class ExtractionRequest(BaseModel):
    """Request to extract knowledge from document."""
    
    document_id: str
    model_id: Optional[str] = None  # Target model for association
    threshold: Optional[float] = Field(None, ge=0, le=1)
    auto_classify: bool = Field(default=True, description="Auto-classify topics")


class ExtractionResult(BaseModel):
    """Result of knowledge extraction."""
    
    document_id: str
    success: bool
    chunks_created: int
    topics_detected: List[str] = Field(default_factory=list)
    confidence_score: Optional[float] = None
    message: str
    processing_time_ms: Optional[int] = None


class DocumentStats(BaseModel):
    """Statistics for documents and knowledge."""
    
    total_documents: int
    by_status: Dict[str, int]
    by_type: Dict[str, int]
    by_model: Dict[str, int]
    total_chunks: int
    total_words: int
    avg_extraction_confidence: Optional[float] = None


class FileUploadResponse(BaseModel):
    """Response for file upload."""
    
    success: bool
    document_id: Optional[str] = None
    file_name: str
    file_size: int
    content_type: DocumentType
    message: str
    status: DocumentStatus
